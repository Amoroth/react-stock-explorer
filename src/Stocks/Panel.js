import React from 'react'
import PropTypes from 'prop-types'
import { withRouter } from 'react-router-dom'

import styles from './Stocks.module.css'

const panel = ({ change, title, short, favorite, onFav, price, history }) => {
  const priceChange = `${change > 0 ? '+' : ''}${(change * 100).toFixed(2)}%`
  let stockName = title
  if (stockName.length > 30) {
    stockName = `${stockName.slice(0, 28)}...`
  }

  const openFullPage = (e) => {
    if ((e.type === 'keypress' && e.which === 13) || e.type === 'click') {
      history.push(`/stock?cmp=${short}`)
    }
  }

  return (
    <div
      className={styles.panel}
      onClick={openFullPage}
      role="button"
      onKeyPress={openFullPage}
      tabIndex={0}
    >
      <div className={styles['panel-upper']}>
        <h6
          className={styles['panel-title']}
          title={stockName[stockName.length - 2] === '.' ? title : null}
        >
          {stockName}
        </h6>
        <button
          className={styles['panel-favorite']}
          onClick={(e) => onFav(e, short)}
          type="button"
        >
          {favorite ? (
            <i className="material-icons" style={{ color: 'red' }}>
              favorite
            </i>
          ) : (
            <i className="material-icons">favorite_border</i>
          )}
        </button>
      </div>
      <div className={styles['panel-details']}>
        <span>{short}</span>
        <span>{`${price} USD`}</span>
        <span style={{ color: change > 0 ? 'green' : 'red' }}>
          {priceChange}
        </span>
      </div>
    </div>
  )
}

panel.propTypes = {
  change: PropTypes.number.isRequired,
  title: PropTypes.string.isRequired,
  short: PropTypes.string.isRequired,
  favorite: PropTypes.bool,
  onFav: PropTypes.func.isRequired,
  history: PropTypes.shape({ push: PropTypes.func.isRequired }).isRequired,
  price: PropTypes.number.isRequired,
}

panel.defaultProps = { favorite: false }

export default withRouter(panel)
