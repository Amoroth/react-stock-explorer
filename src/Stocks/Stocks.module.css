.container {
  margin: auto;
  width: 60%;
  max-width: 800px;
}

.markets-container {
  display: flex;
  flex-wrap: nowrap;
  justify-content: space-evenly;
  margin-bottom: .5em;
}

.stocks-container {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
}

.panel {
  background-color: #fff;
  border-radius: 5px;
  box-shadow: 2px 4px 10px rgba(0,0,0,.0125);
  font-size: 15px;
  margin: 0 0 2% 0;
  padding: 16px;
  transition: box-shadow .1s ease-in-out;
  width: 49%;
  cursor: pointer;
  user-select: none;
}

.panel:hover {
  box-shadow: 2px 4px 10px rgba(0,0,0,.1);
}

.panel-details {
  display: flex;
  justify-content: space-between;
}

.panel-title {
  font-size: 21px;
  font-weight: bold;
  margin: 0 0 10px 0;
}

.panel-upper {
  display: flex;
  justify-content: space-between;
}

.panel-favorite {
  background-color: transparent;
  border: none;
}

.no-favorites {
  background-color: #fff;
  border-radius: 5px;
  box-shadow: 2px 4px 10px rgba(0,0,0,.0125);
  font-size: 15px;
  margin: auto;
  padding: 16px 32px;
  width: 97%;
  text-align: center;
}

@media screen and (max-width: 620px) {
  .container {
    width: 100%;
    flex-direction: column;
    align-items: center;
  }
  
  .stocks-container {
    justify-content: center;
  }

  .panel {
    width: 97%;
  }
}