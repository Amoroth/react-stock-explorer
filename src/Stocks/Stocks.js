import React from 'react'

import Panel from './Panel'
import Spinner from '../shared/Spinner'
import styles from './Stocks.module.css'

class Stocks extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      stocks: [
        {
          companyName: 'Placeholder',
          symbol: 'none',
          close: 1098.77,
          change: 0.0467,
          favorite: false
        }
      ]
    }
  }

  componentDidMount() {
    fetch('https://api.iextrading.com/1.0/stock/market/list/infocus').then((res) => {
      return res.json()
    }).then((json) => {
      this.setState({
        stocks: json
      })
    })

    this.loadFavorites()
  }

  loadFavorites = () => {
    console.log('loading favorites')

    // fetch('https://api.iextrading.com/1.0/stock/market/batch?types=previous&symbols=aapl,msft,googl')
  }

  onFavorite = (event, symbol) => {
    event.stopPropagation()

    console.log('added to favorites')

    // let newStocks = this.state.stocks.slice()
    // newStocks.forEach(element => {
    //   if (element.short === symbol)
    //     element.favorite = !element.favorite
    // })
    // this.setState({
    //   stocks: newStocks
    // }, () => {
    //   localStorage.setItem('stock-explorer_favorites', this.state.stocks.reduce((arr, val) => {
    //     if (val.favorite)
    //       arr.push(val.symbol)
    //     return arr
    //   }, []))
    // })
  }

  render() {
    const stocks = this.state.stocks.map((val) => {
      return <Panel 
        title={val.companyName} 
        short={val.symbol} 
        price={val.close} 
        change={val.changePercent} 
        // favorite={val.favorite}
        key={val.symbol}
        onFav={this.onFavorite}
      />
    })
    return (
      <div className={styles['container']}>
        {this.props.children}
        {this.state.stocks[0].symbol === 'none' ? <Spinner /> : stocks}
      </div>
    )
  }
}

export default Stocks