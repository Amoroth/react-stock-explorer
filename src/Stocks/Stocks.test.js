import React from 'react'
import { shallow } from 'enzyme'

import Stocks from './Stocks'

describe('<Stocks />', () => {
  let props
  let wrapper

  beforeEach(() => {
    props = {
      favorites: ['OTH', 'ZTM', 'TBC'],
      onFavorite: jest.fn(),
      currency: 'USD',
      currencyFormat: jest.fn(),
    }
    wrapper = shallow(<Stocks {...props} />)

    wrapper.setState({
      stocks: [
        { companyName: 'Company', symbol: 'COM', close: 1, changePercent: 0.01 },
        { companyName: 'Other', symbol: 'OTH', close: 2, changePercent: 0.02 },
        { companyName: 'Other 3', symbol: 'OTH 3', close: 3, changePercent: 0.03 },
        { companyName: 'Other 4', symbol: 'OTH 4', close: 4, changePercent: 0.04 },
      ],
    })
  })

  it('renders', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('loads other market', () => {
    const lastButton = wrapper.find('.markets-container').childAt(3)
    const linkname = lastButton.childAt(0).text().toLowerCase()
    lastButton.prop('func')()

    expect(fetch.mock.calls[fetch.mock.calls.length - 1][0]).toBe(`https://api.iextrading.com/1.0/stock/market/list/${linkname}`)
  })

  it('loads favorites', () => {
    const favButton = wrapper.find('.markets-container').childAt(1)
    const favFn = jest.fn()
    wrapper.instance().loadFavorites = favFn
    favButton.prop('func')()
    wrapper.setState({
      stocks: [
        { companyName: 'Some car company', symbol: 'ZTM', close: 1, changePercent: 0.01 },
        { companyName: 'The Burning', symbol: 'TBC', close: 2, changePercent: 0.02 },
        { companyName: 'Other', symbol: 'OTH', close: 3, changePercent: 0.03 },
      ],
      market: 'favorites',
    })

    expect(favFn).toBeCalled()
    expect(wrapper.find('.stocks-container').children().length).toBe(3)
  })

  it('shows spinner if no stocks', () => {
    expect(wrapper.find('withRouter(panel)').exists()).toBe(true)
    expect(wrapper.find('spinner').exists()).toBe(false)

    wrapper.setState({ stocks: [] })

    expect(wrapper.find('withRouter(panel)').exists()).toBe(false)
    expect(wrapper.find('spinner').exists()).toBe(true)
  })
})
