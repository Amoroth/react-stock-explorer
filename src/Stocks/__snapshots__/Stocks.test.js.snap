// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Stocks /> renders 1`] = `
<div
  className="container"
>
  <div
    className="markets-container"
  >
    <circularButton
      func={[Function]}
      key="infocus"
    >
      In Focus
    </circularButton>
    <circularButton
      func={[Function]}
      key="favorites"
    >
      Favorites
    </circularButton>
    <circularButton
      func={[Function]}
      key="gainers"
    >
      Gainers
    </circularButton>
    <circularButton
      func={[Function]}
      key="losers"
    >
      Losers
    </circularButton>
  </div>
  <div
    className="stocks-container"
  >
    <withRouter(panel)
      change={0.01}
      currency="USD"
      favorite={false}
      key="COM"
      onFav={[MockFunction]}
      short="COM"
      title="Company"
    />
    <withRouter(panel)
      change={0.02}
      currency="USD"
      favorite={true}
      key="OTH"
      onFav={[MockFunction]}
      short="OTH"
      title="Other"
    />
    <withRouter(panel)
      change={0.03}
      currency="USD"
      favorite={false}
      key="OTH 3"
      onFav={[MockFunction]}
      short="OTH 3"
      title="Other 3"
    />
    <withRouter(panel)
      change={0.04}
      currency="USD"
      favorite={false}
      key="OTH 4"
      onFav={[MockFunction]}
      short="OTH 4"
      title="Other 4"
    />
  </div>
</div>
`;
